src/workerd/jsg/modules.c++

```diff
@@ -4,10 +4,38 @@
 
 #include "modules.h"
 #include "promise.h"
+#include <kj/mutex.h>
 
 namespace workerd::jsg {
 namespace {
 
+class CompileCache {
+  // The CompileCache is used to hold cached compilation data for built-in JavaScript modules.
+public:
+  void add(kj::StringPtr key, std::unique_ptr<v8::ScriptCompiler::CachedData> cached) const {
+    cache.lockExclusive()->insert(kj::str(key), kj::mv(cached));
+  }
+
+  kj::Maybe<v8::ScriptCompiler::CachedData&> find(kj::StringPtr key) const {
```

Comment:
It's worth noting that the memory-safety of this relies on the assumption that entries are never, ever removed from the cache (including never being replaced). I think that's accurate but it's probably worth calling out.

If we ever changed things so that we remove entries from the cache then we'd have to make this return an atomic refcount or some such.

Reply:
I think the only reason why we would need to replace the item in the cache is if some configuration detail on the isolate was changed, or if different isolates used different configuration options that impacted the compilation from one isolate to another. Node.js, for instance, implements a cache-refresh option that is used when the process is started with snapshot data (https://github.com/nodejs/node/blob/265ea1e74ef429f7c27f05ac4cc9136adf2e8d9b/src/node.cc#L1187)

Reply:
Added a comment.

Reply:
I think we would just have the cache key bake in any options that affect the compilation. We wouldn't want the cache to flap back and forth between different option sets as isolates with differing options get loaded.